[12.0] rest connector

robinkeunen opened the pull request:
The flow to take a share is the following, the connector must communicate with an external easy_my_coop API to keep sync. The steps are:

- [x] fetch subscription requests (SR)
- [x] Validating a SR
  - [x] For a new cooperator
  - [x] For an existing cooperator (capital increase)
- [x] Record payment aka release capital
- [x] Creates a share line.

When merging, watch out for :
- https://github.com/coopiteasy/investor-wallet-platform/pull/73?expand=1
- https://github.com/coopiteasy/investor-wallet-platform/pull/72

robinkeunen committed:
[ADD] easy_my_coop_connector

robinkeunen committed:
[IMP] emcc: catch 403, 404, 500 return code

robinkeunen committed:
[ADD] emcc: manual import wizard

robinkeunen committed:
[ADD] emcc: cron to fetch requests

robinkeunen committed:
[ADD] emcc: send validate request to platform

robinkeunen committed:
[IMP] emca: sr_service.validate returns the new invoice

robinkeunen committed:
[REF] emcc: abstract adapter

robinkeunen committed:
[ADD] emcc: post payment for emc-api invoice sends the payment to platform

robinkeunen committed:
[REF] emcc: components to their own directory

robinkeunen committed:
[FIX] emcc: name in backend view

robinkeunen committed:
[IMP] emca: timestamp exported data

robinkeunen committed:
[ADD] emca: add form pages for API fields

robinkeunen committed:
[ADD] emcc: account account bindings

robinkeunen committed:
[IMP] emcc: add logs

robinkeunen committed:
[FIX] emca: return invoice number innstead of name

robinkeunen committed:
[IMP] emca: add validate route

robinkeunen committed:
[FIX] emca: do not copy external_id

robinkeunen committed:
[IMP] emcc: default date to import wizard

robinkeunen committed:
[FIX] emca: patch for json

vvrossem committed:
[ADD] emc_connector: ACL for easy_my_coop.easy_my_coop_manager

robinkeunen commented on easy_my_coop_api/controllers/controllers.py:
```diff
@@ -3,11 +3,30 @@
 from odoo.addons.base_rest.controllers import main
 
 
+def patch_for_json(path_re):
```
@vvrossem @houssine78 I stumbled in this, it might be related to the bug on the test server.

enricostano commented:
cc/ @cesarlr @danypr92

robinkeunen committed:
[REF] pre-commit semanticless fixes

robinkeunen committed:
[REF] split coop.py

houssine78 committed:
[FIX] fix url join with better lib

houssine78 committed:
[FIX] quick and dirty fix for url join

robinkeunen committed:
precommit

victor-champonnois committed:
[IMP] easy_my_coop: don't display name on subscription request

Use firstname and lastname instead

victor-champonnois committed:
[ADD] easy_my_coop:  migration script : fill first/lastnames

robinkeunen committed:
[FIX] emc_api: wrong call to _process_method

robinkeunen committed:
[FIX] easy_my_coop: missing prefix in demo data

robinkeunen committed:
[FIX] emc_api: validate accepts waiting state

robinkeunen committed:
[ADD] emc_api: approval fields in subscription request services

robinkeunen committed:
[ADD] emc_connector: approval fields on subscription request adapter

robinkeunen committed:
[FIX] emc_api: decrement external api id trial counter

robinkeunen committed:
[IMP] emc_api,emc_connector: binding demo data

robinkeunen committed:
[IMP] emc_api: restrict access to records managed by api clients

1st try: add an api user group and two ir.rules
- for api user group, domain: (1 = 1)
- for emc_user_group, domain: (external_id_generated = False)

This solution does not work because any group ir.rule bypasses
these rules, most notably, the rules from iwp to restrict access
to own structure.

This solution adds a global rules: if records were fetched or created
by actions resulting from api calls, no one can update or delete them.

The api bypasses these rules with .sudo.

A major drawback of this solution is that only Administrator will
be able to manipulate these records in the future. This may or may
not be good thing.

victor-champonnois committed:
[FIX] easy_my_coop: demo data for required firstname + lastname

[FIX] emc: fix demo data and tests for firstname

victor-champonnois committed:
[FIX] easy_my_coop: use firstname and lastname instead of name

- firstname in mail templates
- lastname in form breadcrumb (record name)

robinkeunen committed:
[REF] easy_my_coop: move send_capital_release_request to invoice

robinkeunen committed:
[REF] easy_my_coop: send waiting list email in function

robinkeunen committed:
[REF] easy_my_coop: all mails are configurable

robinkeunen committed:
[REF] easy_my_coop: extract send_confirmation_email

robinkeunen committed:
[REF] emc: consistent mail functions names

robinkeunen committed:
[IMP] emc_connector: cascade on bounded record deletion

robinkeunen committed:
[IMP] emc_api: update manages state change through methods

robinkeunen committed:
[REF] emc_api: remove validate route

robinkeunen committed:
[REF] emc_connector: remove validate route usages

robinkeunen committed:
[REF] emc_connector: split adapters into seperate files

robinkeunen commented:
Hi @enricostano and @danypr92 , can you review the changes to emc_api? It might affect some of your processes.

robinkeunen commented:
@huguesdk @remytms this is a bit heavy to review completely but a review of changes to `easy_my_coop` module would be very useful.

robinkeunen committed:
[FIX] tests

robinkeunen committed:
[IMP] emc_connector: send state changes to api

robinkeunen committed:
[ADD] emc_api: optional request fields

robinkeunen committed:
[ADD] emc_connector: optional request fields

huguesdk commented on easy_my_coop/models/operation_request.py:
```diff
@@ -282,23 +282,25 @@ def validate(self):
                 )
 
-    def get_share_trans_mail_template(self):
+    def _get_share_transfert_mail_template(self):
```
why is there a separate method for each type of message, while the contents are basically the same? it could be one method, used with arguments. the correspondence between all of the data (type of data, boolean field that controls whether to send, and template) could be defined as "constants".

huguesdk commented on easy_my_coop_connector/tests/test_payment.py:
```diff
@@ -0,0 +1,70 @@
+
+    def test_post_payment_sends_and_binds_request(self):
+        srequest = self.browse_ref("easy_my_coop.subscription_request_1_demo")
+        with patch.object(requests, "post") as mock_get:
```
do you know about [`requests_mock`](https://pypi.org/project/requests-mock/)?

robinkeunen replied:
no, I'll take a look next time :-)

huguesdk commented on easy_my_coop_connector/models/subscription_request.py:
```diff
@@ -0,0 +1,142 @@
+            sr_adapter = SubscriptionRequestAdapter(backend=backend, record=self)
+            _, request_dict = sr_adapter.update({"state": "done"})
+
+        return invoice
+
+    @api.multi
+    def block_subscription_request(self):
+        self.ensure_one()
+        super().block_subscription_request()
+        if self.source == "emc_api":
+            backend = self.env["emc.backend"].get_backend()
+            sr_adapter = SubscriptionRequestAdapter(backend=backend, record=self)
+            _, request_dict = sr_adapter.update({"state": "block"})
+
+    @api.multi
+    def unblock_subscription_request(self):
+        self.ensure_one()
+        super().unblock_subscription_request()
+        if self.source == "emc_api":
+            backend = self.env["emc.backend"].get_backend()
+            sr_adapter = SubscriptionRequestAdapter(backend=backend, record=self)
+            _, request_dict = sr_adapter.update({"state": "draft"})
+
+    @api.multi
+    def cancel_subscription_request(self):
+        self.ensure_one()
+        super().cancel_subscription_request()
+        if self.source == "emc_api":
+            backend = self.env["emc.backend"].get_backend()
+            sr_adapter = SubscriptionRequestAdapter(backend=backend, record=self)
+            _, request_dict = sr_adapter.update({"state": "cancelled"})
+
+    @api.multi
+    def put_on_waiting_list(self):
+        self.ensure_one()
+        super().put_on_waiting_list()
+        if self.source == "emc_api":
+            backend = self.env["emc.backend"].get_backend()
+            sr_adapter = SubscriptionRequestAdapter(backend=backend, record=self)
+            _, request_dict = sr_adapter.update({"state": "waiting"})
```
same remark as earlier: could all these methods be refactored into one?

huguesdk reviewed (commented):
just a few comments, in addition to those in the merged pull requests: #190, #193, #194, #195

robinkeunen committed:
[REF] easy_my_coop: typo in word transfer*t*x

Not writing a migration script for field send_share_transfer_email
because it is not used yet in production.

robinkeunen committed:
[REF] easy_my_coop: consistent function names for sending mail

robinkeunen committed:
[REF] easy_my_coop: comment sudo use

robinkeunen committed:
fixup! [IMP] emc_api: restrict access to records managed by api clients

robinkeunen committed:
[IMP] emc_api: use search instead of filtered

Filtered will first ask the database for all subscription requests,
then filter them in python, which is much slower.
A search call will be translated to an sql query that will only
return the matching records.

robinkeunen committed:
[REF] easy_my_coop: extra quote in string

robinkeunen committed:
[IMP] easy_my_coop_connector: use put for updating records

+ moved update function from subscription request to abstract adapter
since it is generic.

robinkeunen committed:
[FIX] emc_connector: do not modify api_dict

Avoid side effect of to_write_values function.

robinkeunen committed:
[ADD] easy_my_coop_api_logs

robinkeunen committed:
[IMP] emc_api: log calls to the api

robinkeunen committed:
[FIX] emc_api: create and update requests based on _external_api_id

robinkeunen committed:
update readme table

robinkeunen committed:
[IMP] emc_api: search only returns draft requests

enricostano commented:
ping @danypr92 @daniquilez @cesarlr

robinkeunen committed:
[IMP] emc_connector: log calls from the connector to the api

robinkeunen committed:
[FIX] emc_connector: url joins and add test

robinkeunen committed:
fixup! [IMP] Don't display name in subscription request

robinkeunen committed:
[IMP] emc_connector, emc_api: remove name

robinkeunen committed:
[IMP] emc_api: add emc_api source

robinkeunen committed:
[REF] emc_api: unused patch_for_json

robinkeunen committed:
[IMP] emc_api: timestamp all exports

robinkeunen committed:
Merge pull request #177 from coopiteasy/12.0-remove-name-from-sub-request-form

[IMP] Don't display name in subscription request

robinkeunen committed:
Merge pull request #196 from coopiteasy/12.0-emc_api-log-operations

[12.0] log succesful calls to the api on platform and customer side

robinkeunen committed:
Merge branch '12.0' into 12.0-emc-rest-connector

Using merge instead of rebase because the divergence is now
too long

cesarlr commented on easy_my_coop/models/subscription_request.py:
```diff
@@ -750,26 +746,37 @@ def validate_subscription_request(self):
     def cancel_subscription_request(self):
         self.ensure_one()
+        if self.state not in ("draft", "waiting", "done", "block"):
+            raise ValidationError(_("You cannot cancel a request in this " "state."))
```
```suggestion
            raise ValidationError(_("You cannot cancel a request in this "
                  "state."
             ))
```
Maybe you wanted to split line to fit 79 chars?

danypr92 commented on easy_my_coop_connector/tests/test_payment.py:
```diff
@@ -0,0 +1,70 @@
+
+    def test_post_payment_sends_and_binds_request(self):
+        srequest = self.browse_ref("easy_my_coop.subscription_request_1_demo")
+        with patch.object(requests, "put") as mock_get:
```
```suggestion
        with patch.object(requests, "put") as mock_put:
```

danypr92 commented on easy_my_coop_connector/tests/test_payment.py:
```diff
@@ -0,0 +1,70 @@
+    def test_post_payment_sends_and_binds_request(self):
+        srequest = self.browse_ref("easy_my_coop.subscription_request_1_demo")
+        with patch.object(requests, "put") as mock_get:
+            mock_get.return_value = mock_response = Mock()
```
```suggestion
            mock_put.return_value = mock_response = Mock()
```

danypr92 commented on easy_my_coop_connector/tests/test_payment.py:
```diff
@@ -0,0 +1,70 @@
+            )
+        )
+
+        with patch.object(requests, "post") as mock_get:
```
```suggestion
        with patch.object(requests, "post") as mock_post:
```

danypr92 commented on easy_my_coop_connector/tests/test_payment.py:
```diff
@@ -0,0 +1,70 @@
+        )
+
+        with patch.object(requests, "post") as mock_get:
+            mock_get.return_value = mock_response = Mock()
```
```suggestion
            mock_post.return_value = mock_response = Mock()
```

danypr92 commented on easy_my_coop_connector/tests/test_subscription_request.py:
```diff
@@ -0,0 +1,142 @@
+
+    def test_validate_request(self):
+        srequest = self.browse_ref("easy_my_coop.subscription_request_1_demo")
+        with patch.object(requests, "put") as mock_get:
```
```suggestion
        with patch.object(requests, "put") as mock_put:
```

danypr92 commented on easy_my_coop/data/mail_template_data.xml:
```diff
@@ -30,7 +30,7 @@
             ><![CDATA[
 <div style="font-family: 'Lucica Grande', Ubuntu, Arial, Verdana, sans-serif; font-size: 12px; color: rgb(34, 34, 34); background-color: #FFF; ">
```
Is the font "Lucica Grande"  "Lucida Grande"?

danypr92 commented on easy_my_coop/models/subscription_request.py:
```diff
@@ -75,10 +76,23 @@ def _check_share_available_to_user(self):
+            )
+            # sudo is needed to change state of invoice linked to a request
+            #  sent through the api
+            mail_template_notif.sudo().send_mail(self.id)
```
Why we need the sudo()?

danypr92 commented on easy_my_coop/models/subscription_request.py:
```diff
@@ -750,26 +746,37 @@ def validate_subscription_request(self):
 
+    def _send_waiting_list_mail(self):
+        if self.company_id.send_waiting_list_email:
+            waiting_list_mail_template = self.env.ref(
```
Can we move this reference to a private method to be overwritten?

danypr92 commented on easy_my_coop_api/models/external_id_mixin.py:
```diff
@@ -26,7 +59,7 @@ def set_external_sequence(self):
         if not sequence:
             sequence = Sequence.sudo().create(
-                {"name": code, "code": code, "number_next": 1}
+                {"name": code, "code": code, "number_next": 100}
```
Why we start in the number 100?

danypr92 commented on easy_my_coop_api/services/account_payment_service.py:
```diff
@@ -74,10 +76,17 @@ def _prepare_create(self, params):
+
+        invoice = {
+            "id": payment.invoice_ids.get_api_external_id(),
+            "name": payment.invoice_ids.number,
```
Can we use "number" instead of "name"?

danypr92 commented on easy_my_coop_api/models/external_id_mixin.py:
```diff
@@ -2,21 +2,54 @@
+    def timestamp_export(self):
+        sudo_self = self.sudo()
+        sudo_self.write({"last_api_export_date": Datetime.now()})
+        sudo_self.search([("first_api_export_date", "=", False)]).write(
```
Is this search looking in the `sudo_self` recordlist or in all the records of model of sudo_self without "first_api_export_date"?

Maybe we can use a filtered in the recordlist to filter the records with "first_api_export_date" set.

danypr92 commented on easy_my_coop_api_logs/security/ir.model.access.csv:
```diff
@@ -0,0 +1,2 @@
+id,name,model_id:id,group_id:id,perm_read,perm_write,perm_create,perm_unlink
+access_emc_api_log_group_system,access_emc_api_log_group_system,easy_my_coop_api_logs.model_emc_api_log,base.group_system,1,0,1,0
```
Is this group the reason to use sudo in the request context?

`base.group_system`
